rules/TOOL_trace_nsiqcppstyle_callbacks.py

```diff
@@ -0,0 +1,80 @@
+from nsiqcppstyle_rulemanager import *
+
+def PrintContextStack(contextStack):
+    if (contextStack != None) and (contextStack.contextstack != None) and (len(contextStack.contextstack) > 0):
+        indent_string = (20 * ' ')
+        print("%sContext stack (%d):" % (indent_string, len(contextStack.contextstack)))
+        for t in contextStack.contextstack:
+            print('%s    %s' % (indent_string, str(t)))
+
+def FunctionScopeRule(lexer, contextStack):
+    print("FunctionScopeRule (lexer, contextStack)")
+    PrintContextStack(contextStack)
+
+def FunctionNameRule(lexer, fullName, decl, contextStack, context):
+    print("FunctionNameRule  (lexer,")
+    print("                   fullName='%s'," % (fullName))
+    print("                   decl='%s'," % (decl))
+    print("                   contextStack,")
+    print("                   context='%s')" % (str(context)))
+    PrintContextStack(contextStack)
+
+
+def PreprocessRule(lexer, contextStack):
+    print("PreprocessRule    (lexer,")
+    print("                   contextStack,")
+    print("                   token=%s" % (lexer.GetCurToken()))
+    PrintContextStack(contextStack)
+
+def CommentRule(lexer, token):
+    print("CommentRule       (lexer, token=%s)" % (lexer.GetCurToken()))
+
+def LineRule(lexer, line, lineNumber):
+    print("--------------------------------------------------")
+    print("LineRule          (lexer, line='%s', lineNumber=%d)" % (line, lineNumber))
+
+def TokenRule(lexer, contextStack):
+    print("TokenRule         (lexer,")
+    print("                   contextStack,")
```

Comment:
Why are you printing just the words `lexer` and `contextStack`?

Reply:
I do not dump the <lexer> object because it is too large, but I want the tool user to see from the output that the rule callback receives a <lexer> object.

I put in "contextStack" for the same reason (i.e., to let the user know he receives a <contextStack> object).  In the case of "contextStack", the stack dump is displayed a few lines below.  I am now planning to remove the "contextStack" placeholder string, moving the actual contextStack dump up into this position:

**Current:**

```
TokenRule         (lexer,
                   contextStack,
                   token=LexToken(LBRACE,'{',6,1,58, False, None))
                   Context stack (1):
                       CLASS_BLOCK, 'foo', LexToken(LBRACE,'{',6,1,58, False, None), LexToken(RBRACE,'}',8,1,80, False, None)
```

**Proposed:**

```
TokenRule         (lexer,
                   Context stack (1):
                       CLASS_BLOCK, 'foo', LexToken(LBRACE,'{',6,1,58, False, None), LexToken(RBRACE,'}',8,1,80, False, None),
                   token=LexToken(LBRACE,'{',6,1,58, False, None))
```

Reply:
This looks much better. Thanks